File: src/entidades/Entidad.java
package entidades;

import utils.GameBoard;
import utils.Posicion;
import utils.Rng;

public abstract class Entidad {
    private static int contadorId = 0;

    protected Posicion posicion;
    protected char simbolo;
    protected int vida;
    protected int vidaMax;
    protected int id;
    protected int danioInfligido;
    protected int danioRecibido;
    protected int kills;

    // Historial de posiciones recientes para evitar ciclos
    private static final int HISTORIAL_MAX = 5;
    private int[] historialFilas = new int[HISTORIAL_MAX];
    private int[] historialCols = new int[HISTORIAL_MAX];
    private int historialSize = 0;
    private int historialIdx = 0;

    // Las 8 direcciones: cardinales + diagonales
    private static final int[][] MOVIMIENTOS = {
        {-1, 0}, {1, 0}, {0, -1}, {0, 1},
        {-1, -1}, {-1, 1}, {1, -1}, {1, 1}
    };

    public Entidad(Posicion pos, char simbolo, int vida) {
        this.id = contadorId++;
        this.posicion = pos;
        this.simbolo = simbolo;
        this.vida = vida;
        this.vidaMax = vida;
    }

    public static void resetContadorId() {
        contadorId = 0;
    }

    public Posicion getPosicion() {
        return posicion;
    }

    public char getSimbolo() {
        return simbolo;
    }

    public int getVida() {
        return vida;
    }

    public void recibirDanio(int danio) {
        vida -= danio;
        if (vida < 0) vida = 0;
    }

    public boolean estaVivo() {
        return vida > 0;
    }

    public int getId() { return id; }
    public int getDanioInfligido() { return danioInfligido; }
    public int getDanioRecibido() { return danioRecibido; }
    public int getKills() { return kills; }
    public void addDanioRecibido(int cantidad) { danioRecibido += cantidad; }

    public abstract void actuar(GameBoard board);

    protected int distancia(Posicion p1, Posicion p2) {
        return Math.abs(p1.getFila() - p2.getFila()) + Math.abs(p1.getColumna() - p2.getColumna());
    }

    protected Entidad buscarCercano(Class<?> tipo, int vision, GameBoard board) {
        Entidad mejor = null;
        int distMin = Integer.MAX_VALUE;
        int miFila = posicion.getFila();
        int miCol = posicion.getColumna();
        for (int df = -vision; df <= vision; df++) {
            for (int dc = -vision; dc <= vision; dc++) {
                int fila = miFila + df;
                int col = miCol + dc;
                if (fila >= 0 && fila < board.getFilas() && col >= 0 && col < board.getColumnas()) {
                    Entidad e = board.getEntidad(fila, col);
                    if (e != null && tipo.isInstance(e)) {
                        int dist = distancia(posicion, e.getPosicion());
                        if (dist < distMin) {
                            distMin = dist;
                            mejor = e;
                        }
                    }
                }
            }
        }
        return mejor;
    }

    protected void moverHacia(Posicion destino, GameBoard board) {
        int[][] movs = copiarMovimientos();
        ordenarPorDistancia(movs, destino, true);
        intentarMovimientos(movs, board);
    }

    protected void moverLejos(Posicion enemigoPos, GameBoard board) {
        int[][] movs = copiarMovimientos();
        ordenarPorDistancia(movs, enemigoPos, false);
        intentarMovimientos(movs, board);
    }

    protected void moverRandom(GameBoard board) {
        int[][] movs = copiarMovimientos();
        // Fisher-Yates shuffle
        for (int i = movs.length - 1; i > 0; i--) {
            int j = Rng.rng.nextInt(i + 1);
            int[] tmp = movs[i];
            movs[i] = movs[j];
            movs[j] = tmp;
        }
        intentarMovimientos(movs, board);
    }

    private int[][] copiarMovimientos() {
        int[][] copia = new int[MOVIMIENTOS.length][2];
        for (int i = 0; i < MOVIMIENTOS.length; i++) {
            copia[i][0] = MOVIMIENTOS[i][0];
            copia[i][1] = MOVIMIENTOS[i][1];
        }
        return copia;
    }

    private void ordenarPorDistancia(int[][] movs, Posicion objetivo, boolean ascendente) {
        for (int i = 0; i < movs.length - 1; i++) {
            int mejorIdx = i;
            int mejorDist = distancia(
                new Posicion(posicion.getFila() + movs[i][0], posicion.getColumna() + movs[i][1]),
                objetivo);
            for (int j = i + 1; j < movs.length; j++) {
                int dist = distancia(
                    new Posicion(posicion.getFila() + movs[j][0], posicion.getColumna() + movs[j][1]),
                    objetivo);
                if (ascendente ? dist < mejorDist : dist > mejorDist) {
                    mejorDist = dist;
                    mejorIdx = j;
                }
            }
            int[] tmp = movs[i];
            movs[i] = movs[mejorIdx];
            movs[mejorIdx] = tmp;
        }
    }

    private boolean estaEnHistorial(int fila, int col) {
        for (int i = 0; i < historialSize; i++) {
            if (historialFilas[i] == fila && historialCols[i] == col) return true;
        }
        return false;
    }

    private boolean intentarMovimientos(int[][] movs, GameBoard board) {
        // Primer paso: intentar todo excepto posiciones del historial
        for (int[] mov : movs) {
            int nf = posicion.getFila() + mov[0];
            int nc = posicion.getColumna() + mov[1];
            if (estaEnHistorial(nf, nc)) continue;
            if (moverSiPosible(nf, nc, board)) {
                return true;
            }
        }
        // Si nada funciono, permitir volver a posiciones del historial
        for (int[] mov : movs) {
            int nf = posicion.getFila() + mov[0];
            int nc = posicion.getColumna() + mov[1];
            if (estaEnHistorial(nf, nc)) {
                if (moverSiPosible(nf, nc, board)) {
                    return true;
                }
            }
        }
        return false;
    }

    private boolean moverSiPosible(int nuevaFila, int nuevaCol, GameBoard board) {
        if (nuevaFila < 0 || nuevaFila >= board.getFilas() || nuevaCol < 0 || nuevaCol >= board.getColumnas()) {
            return false;
        }

        // Aliados detectan trampas y las esquivan
        if (this instanceof Aliado && board.getTrampa(nuevaFila, nuevaCol) != null) {
            return false;
        }

        Entidad destino = board.getEntidad(nuevaFila, nuevaCol);

        // Enemigo ataca Aliado
        if (this instanceof Enemigo && destino instanceof Aliado) {
            Aliado aliado = (Aliado) destino;
            if (aliado.getTurnosInvencible() > 0) {
                // Aliado invencible: el enemigo muere
                aliado.danioInfligido += this.vida;
                this.danioRecibido += this.vida;
                aliado.kills++;
                this.recibirDanio(this.vida);
                board.setEntidad(posicion.getFila(), posicion.getColumna(), null);
                return false;
            }
            int danioEnemigo = ((Enemigo) this).getDanio();
            this.danioInfligido += danioEnemigo;
            aliado.danioRecibido += danioEnemigo;
            aliado.recibirDanio(danioEnemigo);
            // Contraataque: dano base configurable + dano extra por arma
            int contraataque = aliado.getDanioBaseMin() + Rng.rng.nextInt(aliado.getDanioBaseMax() - aliado.getDanioBaseMin() + 1) + aliado.getDanioExtra();
            aliado.danioInfligido += contraataque;
            this.danioRecibido += contraataque;
            this.recibirDanio(contraataque);
            if (!this.estaVivo()) {
                aliado.kills++;
                board.setEntidad(posicion.getFila(), posicion.getColumna(), null);
                return false;
            }
            if (aliado.estaVivo()) {
                return false;
            }
            this.kills++;
            board.setEntidad(nuevaFila, nuevaCol, null);
        }

        // Aliado con estrella ataca Enemigo
        if (this instanceof Aliado && destino instanceof Enemigo) {
            Aliado aliado = (Aliado) this;
            if (aliado.getTurnosInvencible() > 0) {
                aliado.danioInfligido += destino.vida;
                destino.danioRecibido += destino.vida;
                aliado.kills++;
                destino.recibirDanio(destino.vida);
                board.setEntidad(nuevaFila, nuevaCol, null);
            }
        }

        if (board.getEntidad(nuevaFila, nuevaCol) != null) {
            return false;
        }

        int filaVieja = posicion.getFila();
        int colVieja = posicion.getColumna();
        board.setEntidad(filaVieja, colVieja, null);
        posicion.setFila(nuevaFila);
        posicion.setColumna(nuevaCol);
        board.setEntidad(nuevaFila, nuevaCol, this);

        // Anadir posicion vieja al historial (circular)
        historialFilas[historialIdx] = filaVieja;
        historialCols[historialIdx] = colVieja;
        historialIdx = (historialIdx + 1) % HISTORIAL_MAX;
        if (historialSize < HISTORIAL_MAX) historialSize++;
        return true;
    }
}

File: web/js/entidad.js
import * as Rng from './rng.js';

const HISTORIAL_MAX = 5;
const MOVIMIENTOS = [
    [-1, 0], [1, 0], [0, -1], [0, 1],
    [-1, -1], [-1, 1], [1, -1], [1, 1]
];

let contadorId = 0;

export function resetContadorId() {
    contadorId = 0;
}

// ==================== Entidad base ====================

export class Entidad {
    constructor(fila, columna, simbolo, vida) {
        this.id = contadorId++;
        this.fila = fila;
        this.columna = columna;
        this.simbolo = simbolo;
        this.vida = vida;
        this.vidaMax = vida;
        this.danioInfligido = 0;
        this.danioRecibido = 0;
        this.kills = 0;

        // Historial circular
        this._historialFilas = new Array(HISTORIAL_MAX).fill(0);
        this._historialCols = new Array(HISTORIAL_MAX).fill(0);
        this._historialSize = 0;
        this._historialIdx = 0;
    }

    recibirDanio(danio) {
        this.vida -= danio;
        if (this.vida < 0) this.vida = 0;
    }

    estaVivo() {
        return this.vida > 0;
    }

    addDanioRecibido(cantidad) {
        this.danioRecibido += cantidad;
    }

    actuar(board) {
        // Override
    }

    distancia(f1, c1, f2, c2) {
        return Math.abs(f1 - f2) + Math.abs(c1 - c2);
    }

    buscarCercano(tipo, vision, board) {
        let mejor = null;
        let distMin = Infinity;
        for (let df = -vision; df <= vision; df++) {
            for (let dc = -vision; dc <= vision; dc++) {
                const f = this.fila + df;
                const c = this.columna + dc;
                if (f >= 0 && f < board.filas && c >= 0 && c < board.columnas) {
                    const e = board.getEntidad(f, c);
                    if (e !== null && e instanceof tipo) {
                        const dist = this.distancia(this.fila, this.columna, e.fila, e.columna);
                        if (dist < distMin) {
                            distMin = dist;
                            mejor = e;
                        }
                    }
                }
            }
        }
        return mejor;
    }

    moverHacia(destF, destC, board) {
        const movs = this._copiarMovimientos();
        this._ordenarPorDistancia(movs, destF, destC, true);
        this._intentarMovimientos(movs, board);
    }

    moverLejos(enemigoF, enemigoC, board) {
        const movs = this._copiarMovimientos();
        this._ordenarPorDistancia(movs, enemigoF, enemigoC, false);
        this._intentarMovimientos(movs, board);
    }

    moverRandom(board) {
        const movs = this._copiarMovimientos();
        // Fisher-Yates shuffle
        for (let i = movs.length - 1; i > 0; i--) {
            const j = Rng.nextInt(i + 1);
            const tmp = movs[i];
            movs[i] = movs[j];
            movs[j] = tmp;
        }
        this._intentarMovimientos(movs, board);
    }

    _copiarMovimientos() {
        return MOVIMIENTOS.map(m => [m[0], m[1]]);
    }

    _ordenarPorDistancia(movs, objF, objC, ascendente) {
        // Selection sort — port exacto del Java
        for (let i = 0; i < movs.length - 1; i++) {
            let mejorIdx = i;
            let mejorDist = this.distancia(this.fila + movs[i][0], this.columna + movs[i][1], objF, objC);
            for (let j = i + 1; j < movs.length; j++) {
                const dist = this.distancia(this.fila + movs[j][0], this.columna + movs[j][1], objF, objC);
                if (ascendente ? dist < mejorDist : dist > mejorDist) {
                    mejorDist = dist;
                    mejorIdx = j;
                }
            }
            const tmp = movs[i];
            movs[i] = movs[mejorIdx];
            movs[mejorIdx] = tmp;
        }
    }

    _estaEnHistorial(fila, col) {
        for (let i = 0; i < this._historialSize; i++) {
            if (this._historialFilas[i] === fila && this._historialCols[i] === col) return true;
        }
        return false;
    }

    _intentarMovimientos(movs, board) {
        // Primer paso: intentar todo excepto posiciones del historial
        for (const mov of movs) {
            const nf = this.fila + mov[0];
            const nc = this.columna + mov[1];
            if (this._estaEnHistorial(nf, nc)) continue;
            if (this._moverSiPosible(nf, nc, board)) return true;
        }
        // Fallback: permitir volver a posiciones del historial
        for (const mov of movs) {
            const nf = this.fila + mov[0];
            const nc = this.columna + mov[1];
            if (this._estaEnHistorial(nf, nc)) {
                if (this._moverSiPosible(nf, nc, board)) return true;
            }
        }
        return false;
    }

    _moverSiPosible(nuevaFila, nuevaCol, board) {
        if (nuevaFila < 0 || nuevaFila >= board.filas || nuevaCol < 0 || nuevaCol >= board.columnas) {
            return false;
        }

        // Celdas vacias (abismo) son intransitables
        if (board.esVacio && board.esVacio(nuevaFila, nuevaCol)) {
            return false;
        }

        // Aliados detectan trampas y las esquivan
        if (this instanceof Aliado && board.getTrampa(nuevaFila, nuevaCol) !== null) {
            return false;
        }

        const destino = board.getEntidad(nuevaFila, nuevaCol);

        // Enemigo ataca Muro destructible (vida < 9999)
        if (this instanceof Enemigo && destino instanceof Muro && destino.vida < 9999) {
            const danioMuro = this.getDanio();
            this.danioInfligido += danioMuro;
            destino.recibirDanio(danioMuro);
            if (!destino.estaVivo()) {
                board.setEntidad(nuevaFila, nuevaCol, null);
            }
            return false; // no se mueve, solo ataca
        }

        // Enemigo ataca Aliado
        if (this instanceof Enemigo && destino instanceof Aliado) {
            const aliado = destino;
            if (aliado.turnosInvencible > 0) {
                // Aliado invencible: el enemigo muere
                aliado.danioInfligido += this.vida;
                this.danioRecibido += this.vida;
                aliado.kills++;
                this.recibirDanio(this.vida);
                board.setEntidad(this.fila, this.columna, null);
                return false;
            }
            const danioEnemigo = this.getDanio();
            this.danioInfligido += danioEnemigo;
            aliado.danioRecibido += danioEnemigo;
            aliado.recibirDanio(danioEnemigo);
            // Contraataque
            const contraataque = aliado.danioBaseMin + Rng.nextInt(aliado.danioBaseMax - aliado.danioBaseMin + 1) + aliado.danioExtra;
            aliado.danioInfligido += contraataque;
            this.danioRecibido += contraataque;
            this.recibirDanio(contraataque);
            if (!this.estaVivo()) {
                aliado.kills++;
                board.setEntidad(this.fila, this.columna, null);
                return false;
            }
            if (aliado.estaVivo()) {
                return false;
            }
            this.kills++;
            board.setEntidad(nuevaFila, nuevaCol, null);
        }

        // Aliado con estrella ataca Enemigo
        if (this instanceof Aliado && destino instanceof Enemigo) {
            const aliado = this;
            if (aliado.turnosInvencible > 0) {
                aliado.danioInfligido += destino.vida;
                destino.danioRecibido += destino.vida;
                aliado.kills++;
                destino.recibirDanio(destino.vida);
                board.setEntidad(nuevaFila, nuevaCol, null);
            }
        }

        if (board.getEntidad(nuevaFila, nuevaCol) !== null) {
            return false;
        }

        const filaVieja = this.fila;
        const colVieja = this.columna;
        board.setEntidad(filaVieja, colVieja, null);
        this.fila = nuevaFila;
        this.columna = nuevaCol;
        board.setEntidad(nuevaFila, nuevaCol, this);

        // Anadir posicion vieja al historial (circular)
        this._historialFilas[this._historialIdx] = filaVieja;
        this._historialCols[this._historialIdx] = colVieja;
        this._historialIdx = (this._historialIdx + 1) % HISTORIAL_MAX;
        if (this._historialSize < HISTORIAL_MAX) this._historialSize++;
        return true;
    }
}

// ==================== Aliado ====================

export class Aliado extends Entidad {
    constructor(fila, columna, vida, danioBaseMin, danioBaseMax, vision) {
        super(fila, columna, 'A', vida);
        this.vision = vision;
        this.danioBaseMin = danioBaseMin;
        this.danioBaseMax = danioBaseMax;
        this.escudo = 0;
        this.danioExtra = 0;
        this.turnosInvencible = 0;
        this.turnosVelocidad = 0;
        this.objetosRecogidosPersonal = 0;
    }

    addEscudo(cantidad) { this.escudo += cantidad; }
    addDanioExtra(cantidad) { this.danioExtra += cantidad; }
    setTurnosInvencible(turnos) { this.turnosInvencible = turnos; }
    setTurnosVelocidad(turnos) { this.turnosVelocidad = turnos; }
    incrementarObjetosRecogidos() { this.objetosRecogidosPersonal++; }

    curar(cantidad) {
        this.vida = Math.min(this.vida + cantidad, this.vidaMax);
    }

    recibirDanio(danio) {
        if (this.turnosInvencible > 0) return;
        if (this.escudo > 0) {
            if (danio <= this.escudo) {
                this.escudo -= danio;
                return;
            } else {
                danio -= this.escudo;
                this.escudo = 0;
            }
        }
        this.vida -= danio;
        if (this.vida < 0) this.vida = 0;
    }

    actuar(board) {
        if (this.turnosVelocidad > 0) this.turnosVelocidad--;

        const movimientos = this.turnosVelocidad > 0 ? 2 : 1;
        for (let m = 0; m < movimientos; m++) {
            if (!this.estaVivo()) break;
            this._realizarMovimiento(board);
        }
    }

    _realizarMovimiento(board) {
        if (this.turnosInvencible > 0) {
            this.turnosInvencible--;
            // Con estrella, perseguir enemigos
            const masCercano = this.buscarCercano(Enemigo, this.vision, board);
            if (masCercano !== null) {
                this.moverHacia(masCercano.fila, masCercano.columna, board);
            } else {
                this.moverRandom(board);
            }
            return;
        }

        const enemigoCerca = this.buscarCercano(Enemigo, this.vision, board);
        const objetoCerca = this._buscarObjetoCercano(board, this.vision);

        if (enemigoCerca === null) {
            if (objetoCerca !== null) {
                this.moverHacia(objetoCerca.fila, objetoCerca.columna, board);
            } else {
                this.moverRandom(board);
            }
        } else {
            // Hay enemigo cerca
            if (objetoCerca !== null && this.distancia(this.fila, this.columna, objetoCerca.fila, objetoCerca.columna) <= 2) {
                const distObjetoEnemigo = this.distancia(objetoCerca.fila, objetoCerca.columna, enemigoCerca.fila, enemigoCerca.columna);
                const distYoEnemigo = this.distancia(this.fila, this.columna, enemigoCerca.fila, enemigoCerca.columna);
                if (distObjetoEnemigo >= distYoEnemigo) {
                    this.moverHacia(objetoCerca.fila, objetoCerca.columna, board);
                } else {
                    this.moverLejos(enemigoCerca.fila, enemigoCerca.columna, board);
                }
            } else {
                this.moverLejos(enemigoCerca.fila, enemigoCerca.columna, board);
            }
        }
    }

    _buscarObjetoCercano(board, vision) {
        let mejor = null;
        let distMin = Infinity;
        for (let df = -vision; df <= vision; df++) {
            for (let dc = -vision; dc <= vision; dc++) {
                const f = this.fila + df;
                const c = this.columna + dc;
                if (f >= 0 && f < board.filas && c >= 0 && c < board.columnas) {
                    if (board.getObjeto(f, c) !== null) {
                        const dist = this.distancia(this.fila, this.columna, f, c);
                        if (dist < distMin) {
                            distMin = dist;
                            mejor = { fila: f, columna: c };
                        }
                    }
                }
            }
        }
        return mejor;
    }
}

// ==================== Enemigo ====================

export class Enemigo extends Entidad {
    constructor(fila, columna, vida, danioMin, danioMax, vision) {
        super(fila, columna, 'X', vida);
        this.danioMin = danioMin;
        this.danioMax = danioMax;
        this.vision = vision;
    }

    getDanio() {
        return this.danioMin + Rng.nextInt(this.danioMax - this.danioMin + 1);
    }

    actuar(board) {
        const masCercano = this.buscarCercano(Aliado, this.vision, board);
        if (masCercano !== null) {
            this.moverHacia(masCercano.fila, masCercano.columna, board);
        } else {
            this.moverRandom(board);
        }
    }
}

// ==================== EnemigoTanque ====================

export class EnemigoTanque extends Enemigo {
    constructor(fila, columna, vida, danioMin, danioMax, vision) {
        super(fila, columna, vida, danioMin, danioMax, vision);
        this.simbolo = 'T';
        this.turnoInterno = 0;
    }

    actuar(board) {
        this.turnoInterno++;
        if (this.turnoInterno % 2 !== 0) return; // solo actua en turnos pares
        super.actuar(board);
    }
}

// ==================== EnemigoRapido ====================

export class EnemigoRapido extends Enemigo {
    constructor(fila, columna, vida, danioMin, danioMax, vision) {
        super(fila, columna, vida, danioMin, danioMax, vision);
        this.simbolo = 'R';
    }

    actuar(board) {
        super.actuar(board);
        if (this.estaVivo()) {
            super.actuar(board);
        }
    }
}

// ==================== Muro ====================

export class Muro extends Entidad {
    constructor(fila, columna, vida = 9999) {
        super(fila, columna, 'M', vida);
    }

    actuar(board) {
        // Los muros no actuan
    }
}
